### PR TITLE
Split message header into packet/payload/mac

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
Message header is a monolith that is hard to track. For example callbacks from the transport layer would only be able to decode the unencrypted part of the header, so passing an incomplete class may be confusing (need to make it easy to know what data is valid and what not)

 #### Summary of Changes
Starts a split into 3 for message headers.
The code evolved quite a bit, so I only split and kept the monolithic 'MessageHeader' struct. In time, I aim to split it more.

working towards #2716

### CODE REVIEW

**Comment by woody-apple** [comment]
@pan-apple ?

**Review by pan-apple — APPROVED** [review]
Changes look good to me. There's a minor nit...

File: .vscode/tasks.json [hunk 1]
```diff
@@ -101,6 +101,19 @@
...
+        {
+            "label": "Build Lighting App (Linux)",
+            "type": "shell",
+            "command": "scripts/examples/gn_build_example.sh examples/lighting-app/linux out/lighting-app",
+            "group": "build",
+            "problemMatcher": {
+                "base": "$gcc",
+                "fileLocation": [
+                    "relative",
+                    "${workspaceFolder}/out/chip_tool/"
```
Comment (mspang) [inline .vscode/tasks.json:113]: chip_tool -> lighting-app

File: src/transport/MessageHeader.cpp [hunk 2]
```diff
@@ -217,52 +211,31 @@ CHIP_ERROR MessageHeader::DecodeEncryptedHeader(const uint8_t * data, size_t siz
...
+CHIP_ERROR PacketHeader::Encode(uint8_t * data, size_t size, size_t * encode_size, Header::Flags payloadFlags) const
 {
     CHIP_ERROR err  = CHIP_NO_ERROR;
     uint8_t * p     = data;
     uint16_t header = kHeaderVersion << kVersionShift;
 
     VerifyOrExit(size >= EncodeSizeBytes(), err = CHIP_ERROR_INVALID_ARGUMENT);
 
+    // Payload flags are restricted.
+    VerifyOrExit((payloadFlags.value & Header::Flags::kVendorIdPresent) == payloadFlags.value, err = CHIP_ERROR_INVALID_ARGUMENT);
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/transport/MessageHeader.cpp:223]: I want to validate that payload flags contains only valid flags (in this case vendor id). So I am saying 'If I restrict payload flags to the values I know are valid, the flags are unchanged'.

So if any other flag is set (likesourceid or destid), then this comparison should fail.

[the remark this answers is not in the document]
Comment (andy31415) [reply src/transport/MessageHeader.cpp]: I added a 'valid payload flags' constant.

**Review by mspang — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Review by rwalker-apple — APPROVED** [review]

**Comment by rwalker-apple** [comment]
@hawk248 @jelderton @saurabhst ?

**Review by jelderton — APPROVED** [review]

**Pull request merged**